kernel/xpmem_pfn.c

```diff
@@ -226,18 +226,103 @@ xpmem_vaddr_to_pte_size(struct mm_struct *mm, u64 vaddr, u64 *size)
 	return pte;
 }
 
+static void
+xpmem_put_page(struct page *page)
+{
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(4, 6, 0)
+	put_page(page);
+#else
+	page_cache_release(page);
+#endif
+}
+
+struct remap_context {
+	struct xpmem_segment *seg;
+	struct page **pages;
+	unsigned long index;
+	unsigned long count;
+	struct vm_area_struct *vma;
+	u64 vaddr;
+	int result;
+};
+
+#ifdef HAVE_LATEST_APPLY_TO_PAGE_RANGE
+static int remap_func(pte_t *pte, unsigned long addr, void *data)
+#else
+static int remap_func(pte_t *pte, pgtable_t token, unsigned long addr,
+		      void *data)
+#endif
+{
+	struct remap_context *ctx = data;
+	struct xpmem_segment *seg = ctx->seg;
+	struct page *page = ctx->pages[ctx->index++];
+	unsigned long pfn = page_to_pfn(page);
+	unsigned long old_pfn = pte_pfn(*pte);
+
+	BUG_ON(!pfn_valid(pfn));
+
+	if (addr == ctx->vaddr) {
+		if (!old_pfn || pfn == old_pfn) {
+			ctx->result = VM_FAULT_NOPAGE;
+		} else {
+			/* should not be possible, but just in case */
+			XPMEM_DEBUG("remap_func: vaddr:%lx pfn mismatch: "
+				    "%ld != %ld\n", addr, old_pfn, pfn);
+		}
+	}
+
+	if (old_pfn) {
+		xpmem_put_page(page);
+		atomic_dec(&seg->tg->n_pinned);
+		atomic_inc(&xpmem_my_part->n_unpinned);
```

Comment:
why increasing n_unpinned?

Reply:
assuming that comment here and below means why tracking number of unpinned, as it was anyways not "long term" pinning.

This is to track pages that were once pinned and to match xpmem_put_page() calls. It also to match xpmem previous code behavior.